src/qibolab/instruments/qm/config/config.py

```diff
@@ -0,0 +1,125 @@
+from dataclasses import dataclass, field
+from typing import Union
+
+from qibolab.components.configs import IqConfig, OscillatorConfig
+from qibolab.pulses import Pulse
+
+from ..components import OpxOutputConfig, QmAcquisitionConfig, QmChannel
+from .devices import *
+from .elements import *
+from .pulses import *
+
+__all__ = ["QmConfig"]
+
+
+@dataclass
+class QmConfig:
+    """Configuration for communicating with the ``QuantumMachinesManager``.
+
+    Contains nested ``dataclass`` objects and is serialized using ``asdict``
+    to be sent to the instrument.
+    """
+
+    version: int = 1
+    controllers: dict[str, Controller] = field(default_factory=dict)
+    octaves: dict[str, Octave] = field(default_factory=dict)
+    elements: dict[str, Element] = field(default_factory=dict)
+    pulses: dict[str, Union[QmPulse, QmAcquisition]] = field(default_factory=dict)
+    waveforms: dict[str, Waveform] = field(default_factory=dict)
+    digital_waveforms: dict = field(
+        default_factory=lambda: {"ON": {"samples": [(1, 0)]}}
+    )
+    integration_weights: dict = field(default_factory=dict)
+    mixers: dict = field(default_factory=dict)
+
+    def add_controller(self, device: str):
+        if device not in self.controllers:
+            self.controllers[device] = Controller()
```

Comment:
> `Controller` is not frozen, and it is plenty of mutable elements (dicts and strings).
> 
> Maybe, if you're not overwriting it, you shouldn't do it silently. At least, return a boolean to tell whether it was already present.
> 
> Btw, which is the reason to avoid overwriting?

With the current design we are registering everything by looping over the sequence and registering only devices and channels that are being used, except DC channels with offsets that are always registerd in order to keep qubits at their sweetspots even when they are not used in the sequence. If we overwrite the controller entry here, we will potentially lose ports that were registered from a previous pulse on the same controller.

Regarding the boolean, I could easily return it, but I think that the other method that is calling this would probably just ignore it. If that's the case, what would be the use of returning?

Reply:
> With the current design we are registering everything by looping over the sequence and registering only devices and channels that are being used, except DC channels with offsets that are always registerd in order to keep qubits at their sweetspots even when they are not used in the sequence. If we overwrite the controller entry here, we will potentially lose ports that were registered from a previous pulse on the same controller.

Ok, but instead of making unneeded calls to `.add_x()`, and ignore them, I'd prefer to deduplicate them ahead of time. It's less surprising.

> Regarding the boolean, I could easily return it, but I think that the other method that is calling this would probably just ignore it. If that's the case, what would be the use of returning?

In principle, because you don't know who your caller is. Otherwise, it's not a modular design, and you reduce part of the advantage of having a separate function.

Reply:
> Ok, but instead of making unneeded calls to `.add_x()`, and ignore them, I'd prefer to deduplicate them ahead of time. It's less surprising.

I would agree, particularly with this point. I was planning to implement directly but it is probably slightly more involved than what I initially thought so I moved to later milestone (#968) for now, as it is fully internal. It is not very complicated though, so I may end up doing earlier.